File: src/cleansia_android/partner-app/src/main/java/cz/cleansia/partner/features/profile/viewmodels/PersonalSectionViewModel.kt
package cz.cleansia.partner.features.profile.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import cz.cleansia.core.snackbar.SnackbarController
import cz.cleansia.partner.core.network.ApiErrorTranslator
import cz.cleansia.core.network.ApiResult
import cz.cleansia.partner.data.profile.ProfileRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class PersonalSectionUiState(
    val isLoading: Boolean = false,
    val isSaving: Boolean = false,
    val employeeId: String = "",
    val firstName: String = "",
    val lastName: String = "",
    val birthDate: String = "",
    val phone: String = "",
    val email: String = "",
    val firstNameError: String? = null,
    val lastNameError: String? = null,
    val error: String? = null,
    val isSaved: Boolean = false,
)

@HiltViewModel
class PersonalSectionViewModel @Inject constructor(
    private val profileRepository: ProfileRepository,
    private val errorTranslator: ApiErrorTranslator,
    private val snackbar: SnackbarController,
) : ViewModel() {

    private val _uiState = MutableStateFlow(PersonalSectionUiState())
    val uiState: StateFlow<PersonalSectionUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    private fun load() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            when (val result = profileRepository.getCurrentEmployee()) {
                is ApiResult.Success -> {
                    val e = result.data
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            employeeId = e.id.orEmpty(),
                            firstName = e.firstName.orEmpty(),
                            lastName = e.lastName.orEmpty(),
                            birthDate = e.birthDate.orEmpty(),
                            phone = e.phoneNumber.orEmpty(),
                            email = e.email.orEmpty(),
                        )
                    }
                }
                is ApiResult.Error -> {
                    _uiState.update { it.copy(isLoading = false) }
                    snackbar.showError(errorTranslator.translate(result.error))
                }
            }
        }
    }

    fun onFirstNameChange(v: String) = _uiState.update { it.copy(firstName = v, firstNameError = null, error = null) }
    fun onLastNameChange(v: String) = _uiState.update { it.copy(lastName = v, lastNameError = null, error = null) }
    fun onBirthDateChange(v: String) = _uiState.update { it.copy(birthDate = v, error = null) }
    fun onPhoneChange(v: String) = _uiState.update { it.copy(phone = v, error = null) }
    fun onEmailChange(v: String) = _uiState.update { it.copy(email = v, error = null) }

    fun save() {
        val state = _uiState.value
        var hasError = false
        if (state.firstName.isBlank()) {
            _uiState.update { it.copy(firstNameError = "First name is required") }
            hasError = true
        }
        if (state.lastName.isBlank()) {
            _uiState.update { it.copy(lastNameError = "Last name is required") }
            hasError = true
        }
        if (hasError) return
        if (state.employeeId.isBlank()) {
            snackbar.showError("Profile not loaded yet")
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, error = null) }
            val result = profileRepository.updatePersonalInfo(
                employeeId = state.employeeId,
                firstName = state.firstName.trim(),
                lastName = state.lastName.trim(),
                birthDate = state.birthDate.takeIf { it.isNotBlank() },
                phone = state.phone.takeIf { it.isNotBlank() },
                email = state.email.takeIf { it.isNotBlank() },
            )
            when (result) {
                is ApiResult.Success -> _uiState.update { it.copy(isSaving = false, isSaved = true) }
                is ApiResult.Error -> {
                    _uiState.update { it.copy(isSaving = false) }
                    snackbar.showError(errorTranslator.translate(result.error))
                }
            }
        }
    }

    fun clearError() = _uiState.update { it.copy(error = null) }
}

File: src/cleansia_ios/CleansiaPartner/Sources/Features/Profile/Personal/PersonalSectionViewModel.swift
import CleansiaCore
import CleansiaPartnerApi
import Combine
import Foundation

@MainActor
final class PersonalSectionViewModel: ViewModel {
    struct Loaded: Equatable {
        var employeeId: String
        var firstName: String
        var lastName: String
        var birthDate: Date?
        var phone: String
        var email: String
        var firstNameError: String?
        var lastNameError: String?
    }

    @Published private(set) var state: UiState<Void> = .loading
    @Published private(set) var action: ActionState = .idle
    @Published var form = Loaded(
        employeeId: "",
        firstName: "",
        lastName: "",
        birthDate: nil,
        phone: "",
        email: ""
    )

    let saved = PassthroughSubject<Void, Never>()

    private let client: PartnerProfileClient
    private let snackbar: SnackbarController
    private let localizer = ApiErrorLocalizer()

    init(client: PartnerProfileClient, snackbar: SnackbarController) {
        self.client = client
        self.snackbar = snackbar
    }

    func load() async {
        state = .loading
        switch await client.getCurrentEmployee() {
        case let .success(employee):
            form = Loaded(
                employeeId: employee.id ?? "",
                firstName: employee.firstName ?? "",
                lastName: employee.lastName ?? "",
                birthDate: employee.birthDate?.wrappedDate,
                phone: employee.phoneNumber ?? "",
                email: employee.email ?? ""
            )
            state = .loaded(())
        case let .failure(error):
            state = .error(error)
            snackbar.showError(localizer.message(for: error))
        }
    }

    func save() async {
        guard !action.isSubmitting else { return }
        form.firstNameError = nil
        form.lastNameError = nil

        var hasError = false
        if form.firstName.isBlank {
            form.firstNameError = L10n.Profile.errorFirstNameRequired
            hasError = true
        }
        if form.lastName.isBlank {
            form.lastNameError = L10n.Profile.errorLastNameRequired
            hasError = true
        }
        if hasError { return }
        guard !form.employeeId.isBlank else {
            snackbar.showError(L10n.Profile.errorProfileNotLoaded)
            return
        }

        action = .submitting
        let command = UpdatePersonalInfoCommand(
            employeeId: form.employeeId,
            firstName: form.firstName.trimmed,
            lastName: form.lastName.trimmed,
            birthDate: OpenAPIDateWithoutTime(wrappedDate: form.birthDate),
            phone: form.phone.trimmedOrNil,
            email: form.email.trimmedOrNil
        )
        switch await client.updatePersonalInfo(command) {
        case .success:
            action = .idle
            saved.send()
        case let .failure(error):
            action = .error(localizer.message(for: error))
            snackbar.showError(localizer.message(for: error))
        }
    }
}
